core/src/main/java/io/grpc/internal/TransportSet.java

```diff
@@ -157,13 +174,13 @@ public void run() {
           if (shutdown) {
             return;
           }
-          ClientTransport newActiveTransport = transportFactory.newClientTransport(
-              server, authority);
+          SocketAddress addr = getNextAddress();
```

Comment:
Done. Also added unit tests. PTAL.
The last two commits failed on the Travis mac build. I couldn't reproduce it on my mac.